src/coronawhy_vt/utils/local_processing.py

```diff
@@ -0,0 +1,375 @@
+import json
+import os
+from typing import Callable
+
+import dateutil.parser as parser
+import numpy as np
+import pandas as pd
+import pylab
+import spacy
+
+# TODO All of these functions open a file and do not close it, putting them in the same class would solve some issues
+# TODO lowercase?
+
+
+def make_ngram_map_and_list(filename: str, min_word_len: int = 6, min_num_tokens: int = 0,
+                            max_num_tokens: int = 4) -> (list, dict):
+    """Given the path of a file of tokens, replace every " " with "_".
+
+    :param filename: The path of the file
+    :param min_word_length: Minimum length of vocab entry to be accepted
+    :param min_num_tokens: Minimum number of tokens in the vocab entry
+    :param max_num_tokens: Maximum number of tokens in the vocab entry
+
+    :return (names, ngram_map): (Original tokens with spaces instead of underscores
+        {names: names_with_space_repaced_by_underscores})
+    """
+    ngram_map = {}
+    names = []
+    all_names = []
+
+    file = open(filename).readlines()
+    for line in file:
+        # TODO this is meant to remove the last newline character but can remove the last character of the last line if
+        # file does not end with newline recommend rstrip()
+        all_names.append(line.rstrip())
+
+    for name in all_names:
+        if len(name) >= min_word_len:
+            if " " in name:
+                if min_num_tokens <= len(name.split(" ")) <= max_num_tokens:
+                    newname = name.replace(" ", "_").lower()
+                    ngram_map[name.lower()] = newname
+                    names.append(newname)
+            else:
+                names.append(name.lower())
+    names = np.unique(names)
+    return names, ngram_map
+
+
+# TODO Pull the next 6 functions into a class?
+# These functions determine what blocks are pulled from the paper for matching
+def title_blocks(paper: dict) -> list:
+    """Retrieve title block of a paper in json form.
+
+    :param paper (json): The paper to be parsed
+
+    :return: Lines of text in the paper title
+    """
+    return [{'text': paper['metadata']['title']}]
+
+
+def abstract_blocks(paper: dict) -> list:
+    """Retrieve abstract block of a paper in json form.
+
+    :param paper: The paper to be parsed
+
+    :return: Lines of text in the paper abstract
+    """
+    return paper['abstract']
+
+
+def body_blocks(paper: dict) -> list:
+    """Retrieve body block of a paper in json form.
+
+    :param paper: The paper to be parsed
+
+    :return: Lines of text in the paper body
+    """
+    return paper['body_text']
+
+
+# TODO the reference nlp is an external dependence, either add into function or make it a class
+nlp = spacy.load('en_core_web_lg')
+
+
+def pull_mentions_lemmatized(
+        paths: list,
+        block_selector: Callable[[dict], list],
+        sec_name: str,
+        words: list,
+        replace_dict: dict = None) -> dict:
```

Comment:
make a new line after `replace_dict: dict = None` but before the closing parentheses `)` to make this a bit more readable